File: packages/dtj-typescript/test/protocol.test.ts
/**
 * Unit tests for protocol.ts
 */

import { test, describe, beforeEach, afterEach } from "node:test";
import assert from "node:assert/strict";
import {
  PROTOCOL_VERSION,
  Cmd,
  Resp,
  DictKind,
  Severity,
  TypeTag,
  encodeFrame,
  decodeFrame,
  encodeHello,
  decodeHelloOk,
  encodeOpenSession,
  createOpenSessionMetadata,
  encodeIntern,
  decodeInternOk,
  encodeAppendEvent,
  decodeAppendEventOk,
  encodeFinishSession,
  encodePing,
  decodeError,
  encodeValue,
  severityFromString,
} from "../src/protocol.js";
import { DTJProtocolError, DTJValueError } from "../src/errors.js";

describe("protocol", () => {
  describe("encodeFrame / decodeFrame", () => {
    test("encodes and decodes a simple frame", () => {
      const body = new Uint8Array([1, 2, 3, 4]);
      const frame = encodeFrame(0x01, body);
      const { opcode, body: decoded } = decodeFrame(frame);
      assert.equal(opcode, 0x01);
      assert.deepEqual(decoded, body);
    });

    test("frame length includes opcode byte", () => {
      const body = new Uint8Array([0xAA]);
      const frame = encodeFrame(0x02, body);
      // Length = 1 (opcode) + 1 (body) = 2
      const view = new DataView(frame.buffer, frame.byteOffset, 4);
      assert.equal(view.getUint32(0, true), 2);
    });

    test("throws on frame too large", () => {
      const largeBody = new Uint8Array(1_048_576); // > MAX_FRAME_SIZE - 1
      assert.throws(() => encodeFrame(0x01, largeBody), DTJProtocolError);
    });

    test("throws on frame too short", () => {
      assert.throws(() => decodeFrame(new Uint8Array([0, 0, 0, 0])), DTJProtocolError);
    });

    test("throws on truncated frame", () => {
      const frame = encodeFrame(0x01, new Uint8Array([1, 2, 3]));
      const truncated = frame.slice(0, frame.length - 1);
      assert.throws(() => decodeFrame(truncated), DTJProtocolError);
    });
  });

  describe("Hello / HelloOk", () => {
    test("encodeHello produces correct frame", () => {
      const frame = encodeHello();
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.HELLO);
      assert.equal(body.length, 4);
      const view = new DataView(body.buffer, body.byteOffset, body.byteLength);
      assert.equal(view.getUint32(0, true), PROTOCOL_VERSION);
    });

    test("decodeHelloOk returns protocol version", () => {
      const body = new Uint8Array(4);
      new DataView(body.buffer).setUint32(0, 1, true);
      assert.equal(decodeHelloOk(body), 1);
    });

    test("decodeHelloOk throws on wrong body length", () => {
      assert.throws(() => decodeHelloOk(new Uint8Array([1, 2])), DTJProtocolError);
    });
  });

  describe("OpenSession metadata", () => {
    test("createOpenSessionMetadata generates valid metadata", () => {
      const meta = createOpenSessionMetadata("test.dtj", "my-service", "1.0.0");
      assert.equal(meta.fileName, "test.dtj");
      assert.equal(meta.producerName, "my-service");
      assert.equal(meta.producerVersion, "1.0.0");
      assert.equal(meta.sessionId.length, 16);
      assert.ok(meta.startUtcUnixMs > 0n);
      assert.ok(meta.monoOriginNs > 0n);
    });

    test("encodeOpenSession produces valid frame", () => {
      const meta = createOpenSessionMetadata("test.dtj", "my-service", "1.0.0");
      const frame = encodeOpenSession(meta);
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.OPEN_SESSION);
      assert.ok(body.length > 0);
    });

    test("encodeOpenSession validates producerName length", () => {
      const meta = createOpenSessionMetadata("test.dtj", "x".repeat(33), "1.0.0");
      assert.throws(() => encodeOpenSession(meta), /producerName must be <= 32 bytes/);
    });

    test("encodeOpenSession validates producerVersion length", () => {
      const meta = createOpenSessionMetadata("test.dtj", "my-service", "x".repeat(17));
      assert.throws(() => encodeOpenSession(meta), /producerVersion must be <= 16 bytes/);
    });

    test("encodeOpenSession validates sessionId length", () => {
      const meta = createOpenSessionMetadata("test.dtj", "my-service", "1.0.0");
      meta.sessionId = new Uint8Array(15);
      assert.throws(() => encodeOpenSession(meta), /sessionId must be 16 bytes/);
    });
  });

  describe("Intern / InternOk", () => {
    test("encodeIntern produces valid frame", () => {
      const frame = encodeIntern(DictKind.DOMAIN, "api");
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.INTERN);
      assert.equal(body[0], DictKind.DOMAIN);
      // name_len = 3
      const view = new DataView(body.buffer, 1, 2);
      assert.equal(view.getUint16(0, true), 3);
      assert.equal(new TextDecoder().decode(body.slice(3)), "api");
    });

    test("encodeIntern throws on name too long", () => {
      assert.throws(() => encodeIntern(DictKind.DOMAIN, "x".repeat(1025)), /name too long/);
    });

    test("decodeInternOk returns dictionary ID", () => {
      const body = new Uint8Array(4);
      new DataView(body.buffer).setUint32(0, 42, true);
      assert.equal(decodeInternOk(body), 42);
    });

    test("decodeInternOk throws on wrong body length", () => {
      assert.throws(() => decodeInternOk(new Uint8Array([1, 2, 3])), DTJProtocolError);
    });
  });

  describe("encodeValue", () => {
    test("encodes boolean", () => {
      const { typeTag, body } = encodeValue(true);
      assert.equal(typeTag, TypeTag.BOOL);
      assert.deepEqual(body, new Uint8Array([1]));

      const { typeTag: t2, body: b2 } = encodeValue(false);
      assert.equal(t2, TypeTag.BOOL);
      assert.deepEqual(b2, new Uint8Array([0]));
    });

    test("encodes bigint as I64", () => {
      const { typeTag, body } = encodeValue(12345n);
      assert.equal(typeTag, TypeTag.I64);
      assert.equal(body.length, 8);
      const view = new DataView(body.buffer);
      assert.equal(view.getBigInt64(0, true), 12345n);
    });

    test("encodes integer number as I64", () => {
      const { typeTag, body } = encodeValue(42);
      assert.equal(typeTag, TypeTag.I64);
      const view = new DataView(body.buffer);
      assert.equal(view.getBigInt64(0, true), 42n);
    });

    test("encodes non-integer number as F64", () => {
      const { typeTag, body } = encodeValue(3.14);
      assert.equal(typeTag, TypeTag.F64);
      const view = new DataView(body.buffer);
      assert.equal(view.getFloat64(0, true), 3.14);
    });

    test("encodes Uint8Array as BYTES", () => {
      const data = new Uint8Array([1, 2, 3, 4]);
      const { typeTag, body } = encodeValue(data);
      assert.equal(typeTag, TypeTag.BYTES);
      assert.equal(body.length, 4 + 4); // length prefix + data
      const view = new DataView(body.buffer);
      assert.equal(view.getUint32(0, true), 4);
      assert.deepEqual(body.slice(4), data);
    });

    test("throws on string (must use dictionary)", () => {
      assert.throws(() => encodeValue("hello"), DTJValueError);
    });

    test("throws on unsupported type", () => {
      assert.throws(() => encodeValue({}), DTJValueError);
      assert.throws(() => encodeValue(null), DTJValueError);
      assert.throws(() => encodeValue(undefined), DTJValueError);
    });

    test("throws on non-finite number", () => {
      assert.throws(() => encodeValue(NaN), DTJValueError);
      assert.throws(() => encodeValue(Infinity), DTJValueError);
      assert.throws(() => encodeValue(-Infinity), DTJValueError);
    });
  });

  describe("AppendEvent", () => {
    test("encodeAppendEvent produces valid frame", () => {
      const frame = encodeAppendEvent(
        1000n,
        1, 2, 3, 4,
        Severity.INFO,
        5,
        TypeTag.I64,
        new Uint8Array([0, 0, 0, 0, 0, 0, 0, 42])
      );
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.APPEND_EVENT);
      // Verify structure: 8 + 16 + 1 + 2 + 4 + 1 + 3 + 8 = 43 bytes body
      assert.equal(body.length, 43);
    });

    test("decodeAppendEventOk returns event sequence", () => {
      const body = new Uint8Array(8);
      new DataView(body.buffer).setBigUint64(0, 123n, true);
      assert.equal(decodeAppendEventOk(body), 123n);
    });

    test("decodeAppendEventOk throws on wrong body length", () => {
      assert.throws(() => decodeAppendEventOk(new Uint8Array([1, 2, 3])), DTJProtocolError);
    });
  });

  describe("FinishSession / Ping", () => {
    test("encodeFinishSession produces empty body frame", () => {
      const frame = encodeFinishSession();
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.FINISH_SESSION);
      assert.equal(body.length, 0);
    });

    test("encodePing produces empty body frame", () => {
      const frame = encodePing();
      const { opcode, body } = decodeFrame(frame);
      assert.equal(opcode, Cmd.PING);
      assert.equal(body.length, 0);
    });
  });

  describe("decodeError", () => {
    test("decodes error message", () => {
      const msg = "Something went wrong";
      const body = new TextEncoder().encode(msg);
      assert.equal(decodeError(body), msg);
    });
  });

  describe("severityFromString", () => {
    test("maps severity strings correctly", () => {
      assert.equal(severityFromString("debug"), Severity.DEBUG);
      assert.equal(severityFromString("info"), Severity.INFO);
      assert.equal(severityFromString("warn"), Severity.WARN);
      assert.equal(severityFromString("error"), Severity.ERROR);
      assert.equal(severityFromString("fatal"), Severity.FATAL);
      assert.equal(severityFromString("DEBUG"), Severity.DEBUG);
      assert.equal(severityFromString("WARN"), Severity.WARN);
    });

    test("throws on invalid severity", () => {
      assert.throws(() => severityFromString("invalid"), DTJValueError);
      assert.throws(() => severityFromString(""), DTJValueError);
    });
  });
});

File: packages/dtj-go/protocol_test.go
package dtj

import (
	"bytes"
	"encoding/binary"
	"testing"
)

func TestEncodeDecodeFrame(t *testing.T) {
	body := []byte{1, 2, 3, 4}
	frame, err := EncodeFrame(0x01, body)
	if err != nil {
		t.Fatalf("EncodeFrame failed: %v", err)
	}

	// Verify frame structure
	if len(frame) < 5 {
		t.Fatalf("frame too short: %d", len(frame))
	}
	length := binary.LittleEndian.Uint32(frame[0:4])
	if length != 5 { // 1 opcode + 4 body
		t.Fatalf("wrong frame length: %d", length)
	}
	if frame[4] != 0x01 {
		t.Fatalf("wrong opcode: 0x%02x", frame[4])
	}
	if !bytes.Equal(frame[5:], body) {
		t.Fatalf("body mismatch")
	}

	// Decode
	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != 0x01 {
		t.Fatalf("decoded opcode mismatch: 0x%02x", decoded.Opcode)
	}
	if !bytes.Equal(decoded.Body, body) {
		t.Fatalf("decoded body mismatch")
	}
}

func TestEncodeDecodeHello(t *testing.T) {
	frame, err := EncodeHello()
	if err != nil {
		t.Fatalf("EncodeHello failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpHello {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}

	version, err := DecodeHelloOk(decoded.Body)
	if err != nil {
		t.Fatalf("DecodeHelloOk failed: %v", err)
	}
	if version != ProtocolVersion {
		t.Fatalf("version mismatch: %d != %d", version, ProtocolVersion)
	}
}

func TestOpenSessionMetadata(t *testing.T) {
	meta, err := NewOpenSessionMetadata("test.dtj", "my-service", "1.0.0")
	if err != nil {
		t.Fatalf("NewOpenSessionMetadata failed: %v", err)
	}

	if meta.FileName != "test.dtj" {
		t.Fatalf("filename mismatch: %s", meta.FileName)
	}
	if meta.ProducerName != "my-service" {
		t.Fatalf("producer name mismatch: %s", meta.ProducerName)
	}
	if meta.ProducerVersion != "1.0.0" {
		t.Fatalf("producer version mismatch: %s", meta.ProducerVersion)
	}
	if meta.SessionID == [16]byte{} {
		t.Fatalf("session ID not generated")
	}
	if meta.StartUtcUnixMs == 0 {
		t.Fatalf("start time not set")
	}
	if meta.MonoOriginNs == 0 {
		t.Fatalf("mono origin not set")
	}
}

func TestOpenSessionMetadataValidation(t *testing.T) {
	// Test producer name too long
	_, err := NewOpenSessionMetadata("test.dtj", string(make([]byte, 33)), "1.0.0")
	if err == nil {
		t.Fatalf("expected error for long producer name")
	}

	// Test producer version too long
	_, err = NewOpenSessionMetadata("test.dtj", "my-service", string(make([]byte, 17)))
	if err == nil {
		t.Fatalf("expected error for long producer version")
	}
}

func TestEncodeOpenSession(t *testing.T) {
	meta, err := NewOpenSessionMetadata("test.dtj", "my-service", "1.0.0")
	if err != nil {
		t.Fatalf("NewOpenSessionMetadata failed: %v", err)
	}

	frame, err := EncodeOpenSession(meta)
	if err != nil {
		t.Fatalf("EncodeOpenSession failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpOpenSession {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}
}

func TestEncodeIntern(t *testing.T) {
	frame, err := EncodeIntern(DictKindDomain, "api")
	if err != nil {
		t.Fatalf("EncodeIntern failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpIntern {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}
	if decoded.Body[0] != DictKindDomain {
		t.Fatalf("wrong dict kind: %d", decoded.Body[0])
	}
}

func TestEncodeInternValidation(t *testing.T) {
	// Test name too long
	_, err := EncodeIntern(DictKindDomain, string(make([]byte, 1025)))
	if err == nil {
		t.Fatalf("expected error for long name")
	}
}

func TestDecodeInternOk(t *testing.T) {
	body := make([]byte, 4)
	binary.LittleEndian.PutUint32(body, 42)
	id, err := DecodeInternOk(body)
	if err != nil {
		t.Fatalf("DecodeInternOk failed: %v", err)
	}
	if id != 42 {
		t.Fatalf("id mismatch: %d", id)
	}
}

func TestEncodeAppendEvent(t *testing.T) {
	frame, err := EncodeAppendEvent(
		1000,
		1, 2, 3, 4,
		SeverityInfo,
		5,
		TypeTagI64,
		int64ToBytes(42),
	)
	if err != nil {
		t.Fatalf("EncodeAppendEvent failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpAppendEvent {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}
}

func TestDecodeAppendEventOk(t *testing.T) {
	body := make([]byte, 8)
	binary.LittleEndian.PutUint64(body, 123)
	seq, err := DecodeAppendEventOk(body)
	if err != nil {
		t.Fatalf("DecodeAppendEventOk failed: %v", err)
	}
	if seq != 123 {
		t.Fatalf("seq mismatch: %d", seq)
	}
}

func TestEncodeFinishSession(t *testing.T) {
	frame, err := EncodeFinishSession()
	if err != nil {
		t.Fatalf("EncodeFinishSession failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpFinishSession {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}
	if len(decoded.Body) != 0 {
		t.Fatalf("body should be empty")
	}
}

func TestEncodePing(t *testing.T) {
	frame, err := EncodePing()
	if err != nil {
		t.Fatalf("EncodePing failed: %v", err)
	}

	r := bytes.NewReader(frame)
	decoded, err := DecodeFrame(r)
	if err != nil {
		t.Fatalf("DecodeFrame failed: %v", err)
	}
	if decoded.Opcode != OpPing {
		t.Fatalf("wrong opcode: 0x%02x", decoded.Opcode)
	}
}

func TestEncodeValue(t *testing.T) {
	tests := []struct {
		name    string
		value   any
		wantTag TypeTag
		wantErr bool
	}{
		{"bool true", true, TypeTagBool, false},
		{"bool false", false, TypeTagBool, false},
		{"int8", int8(42), TypeTagI64, false},
		{"int16", int16(42), TypeTagI64, false},
		{"int32", int32(42), TypeTagI64, false},
		{"int64", int64(42), TypeTagI64, false},
		{"int", 42, TypeTagI64, false},
		{"uint8", uint8(42), TypeTagU64, false},
		{"uint16", uint16(42), TypeTagU64, false},
		{"uint32", uint32(42), TypeTagU64, false},
		{"uint64", uint64(42), TypeTagU64, false},
		{"uint", uint(42), TypeTagU64, false},
		{"float32", float32(3.14), TypeTagF64, false},
		{"float64", 3.14, TypeTagF64, false},
		{"bytes", []byte{1, 2, 3}, TypeTagBytes, false},
		{"string", "hello", 0, true}, // strings must be interned
		{"unsupported", struct{}{}, 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			tag, body, err := EncodeValue(tt.value)
			if tt.wantErr {
				if err == nil {
					t.Fatalf("expected error for %s", tt.name)
				}
				return
			}
			if err != nil {
				t.Fatalf("EncodeValue failed for %s: %v", tt.name, err)
			}
			if tag != tt.wantTag {
				t.Fatalf("tag mismatch for %s: 0x%02x != 0x%02x", tt.name, tag, tt.wantTag)
			}
			if len(body) == 0 {
				t.Fatalf("empty body for %s", tt.name)
			}
		})
	}
}

func TestSeverityFromString(t *testing.T) {
	tests := []struct {
		input   string
		want    Severity
		wantErr bool
	}{
		{"debug", SeverityDebug, false},
		{"info", SeverityInfo, false},
		{"warn", SeverityWarn, false},
		{"error", SeverityError, false},
		{"fatal", SeverityFatal, false},
		{"invalid", 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			got, err := SeverityFromString(tt.input)
			if tt.wantErr {
				if err == nil {
					t.Fatalf("expected error for %s", tt.input)
				}
				return
			}
			if err != nil {
				t.Fatalf("SeverityFromString failed for %s: %v", tt.input, err)
			}
			if got != tt.want {
				t.Fatalf("severity mismatch for %s: %d != %d", tt.input, got, tt.want)
			}
		})
	}
}

func TestSeverityString(t *testing.T) {
	tests := []struct {
		sev  Severity
		want string
	}{
		{SeverityDebug, "debug"},
		{SeverityInfo, "info"},
		{SeverityWarn, "warn"},
		{SeverityError, "error"},
		{SeverityFatal, "fatal"},
		{Severity(99), "unknown"},
	}

	for _, tt := range tests {
		if tt.sev.String() != tt.want {
			t.Fatalf("Severity.String() mismatch for %d: %s != %s", tt.sev, tt.sev.String(), tt.want)
		}
	}
}
